### ISSUE TITLE
TF 2.11.0: tf.image.ssim return_index_map=True outputs wrong shape

### ISSUE THREAD

**Issue by sapphire008** [tensorflow/tensorflow#59067]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Have you reproduced the bug with TF nightly?

No

### Source

source

### Tensorflow Version

2.11.0

### Custom Code

Yes

### OS Platform and Distribution

MacOS 13.1

### Python version

3.10.6

### Current Behaviour?

```shell
The documentation for `tf.image.ssim` claims to output

> ... a tensor containing an SSIM value for each pixel for each image in batch if return_index_map is True

However, the output image is smaller than the source image (see example below).

Upon comparing with a more well-known library implemented in PyTorch, I believe the reason for such discrepency is due to the Conv2D padding used.

In PyTorch's implementation, a "SAME" padding is used

https://github.com/Po-Hsun-Su/pytorch-ssim/blob/3add4532d3f633316cba235da1c69e90f0dfb952/pytorch_ssim/__init__.py#L25

However, in the current Tensorflow implementation, "VALID" padding is used

https://github.com/tensorflow/tensorflow/blob/d5b57ca93e506df258271ea00fc29cf98383a374/tensorflow/python/ops/image_ops_impl.py#L4340

Please verify if this is the case.
```

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf # tf.__version__ == "2.11.0"
# B x T x n_mel
shape = (16, 2106, 80, 1)
image1 = np.arange(np.prod(shape))
image1 = (image1 / np.max(image1)) * 10 + 100
image1 = np.reshape(image1, shape)
image2 = np.linspace(0, 1, np.prod(shape))
image2 = np.exp(image2)
image2 = (image2 / np.max(image2)) * 10 + 100
image2 = np.reshape(image2, shape)

out_tf = tf.image.ssim(image1, image2, max_val=255, return_index_map=True)

```
which outputs a tensor of shape [16, 2096, 70].

### Relevant log output

_No response_</details>

**Comment by tilakrayal** [comment]
@sapphire008,
I was facing a different issue/error while executing the mentioned code. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/a18c78fc2f901e1c2c54facba4f6548c/untitled829.ipynb) and provide the dependencies. Thank you!

**Comment by sapphire008** [comment]
@tilakrayal This is an issue with the latest release of tensorflow 2.11.0 as indicated in the original post under "Tensorflow Version". The colab version is still using 2.9.2.

**Comment by tilakrayal** [comment]
@sapphire008,
Thank you for the issue.  In tf.image.ssim when we are using **return_index_map = True**, returns the **index map**; where as in the case of  **return_index_map = False**, it should return the reduced **global value** which was working as expected. 

**O/p for return_index_map = True** :

```
2.12.0-dev20230201
(2695680,)
(2695680,)
(16, 2106, 80, 1)
(2695680,)
(2695680,)
(2695680,)
(16, 2106, 80, 1)
(16,)
```

**O/p for return_index_map = False** :

```
2.12.0-dev20230201
(2695680,)
(2695680,)
(16, 2106, 80, 1)
(2695680,)
(2695680,)
(2695680,)
(16, 2106, 80, 1)
(16, 2096, 70)
```
A tensor containing an SSIM value for each image in batch or a tensor containing an SSIM value for each pixel for each image in batch if **return_index_map** is True. Returned SSIM values are in range (-1, 1], when pixel values are non-negative. Returns a tensor with shape:
```
    broadcast(img1.shape[:-3], img2.shape[:-3]) or broadcast(img1.shape[:-1],
    img2.shape[:-1]).
```

Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/9af60d2910d739bfc1230eb296c8599c/untitled917.ipynb) and also please have a look  at the reference where it is explicitly described that the global ssim value is the mean of the local ssim value: https://medium.com/srm-mic/all-about-structural-similarity-index-ssim-theory-code-in-pytorch-6551b455541e .

**Comment by sapphire008** [comment]
@tilakrayal I believe in your response has the `return_index_map =True` and `return_index_map =False` cases switched. However, the [gist](https://colab.research.google.com/gist/tilakrayal/9af60d2910d739bfc1230eb296c8599c/untitled917.ipynb) version correctly reproduced the problem I have described in the original post. Regardless, the output shape is (16, 2096, 70) when setting `return_index_map=True`, which is not consistent with the input shape of (16, 2106, 80). This is unexpected, because a index map of SSIM calculation is the scores corresponding to the original input, thus their shape should have been consistent. You can also find in my original post for a link to another implementation made in PyTorch, which do produce index map with shapes consistent with the inputs, and has been cited / reproduced by various repositories such as https://github.com/MoonInTheRiver/DiffSinger

**Comment by gaikwadrahul8** [comment]
Hi, @sapphire008 

Apologize for the delayed response and I was able to replicate the same issue and I have executed the same code multiple times and It seems like working as expected with `TF2.11` and I also tried to with our latest pre-release `tensorflow==2.12.0rc1 `and `tensorflow==2.12.0rc0` for your reference I have added [gist-file](https://colab.research.google.com/gist/gaikwadrahul8/a332efd18ba5adb3e34a95ffa4abea92/-59067-test.ipynb) and I completely understood your point here the implementation in PyTorch with "`SAME`" padding  but in Tensorflow we are using "`VALID`" padding, If have I missed something here please let me know ? Thank you!

**Comment by EdwardCuiPeacock** [comment]
@gaikwadrahul8 Is there any reason why `VALID` padding is chosen instead of `SAME` padding in this specific implementation? The returned `index_map` is usually expected as the same size as the inputs, to my understanding. It's also how I see it could be used as a loss function.

**Comment by gaikwadrahul8** [comment]
Hi, @SuryanarayanaY 

Could you please look into this issue ? Thank you!

**Comment by SuryanarayanaY** [comment]
@sapphire008 ,

The shape mismatch might be due to the padding as you noticed. As per the documentation the API the output is:
`Returns a tensor with shape: broadcast(img1.shape[:-3], img2.shape[:-3]) or broadcast(img1.shape[:-1], img2.shape[:-1])`.
 
But for our example the input is of shape (16, 2106, 80, 1) the API returns  output of shape (16, 2096, 70) which seems incorrect and the root cause is due to the padding='VALID'

I have tested the code by changing padding='SAME' and in this case the output is (16, 2106, 80).

All these details captured in attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/d75ef4fcbd69b5211975fd4d29490084/59067_r1.ipynb#scrollTo=G24eYEw6Qi9Q).

It seems a bug for me and I am going to raise a PR for this. Thanks for reporting.

**Comment by kiransair** [comment]
Hi @sapphire008, Thanks for your patience. I have tried by manually adding the padding to input that is being passed to the tf.image.ssim. I got the same shape as the input. please refer to this [gist](https://gist.github.com/kiransair/2bccc1c67bdd0868c8b814e7c376a479). Thank You.
